### ISSUE TITLE
Convert cactbot to use Webpack

### ISSUE THREAD

**Issue by quisquous** [quisquous/cactbot#1869]
I'd like to switch cactbot over to using [Webpack](https://webpack.js.org/).  In other words, running a bundling step prior to using any local files.  For end users, there should be no difference, but this will require developers to change their workflow.  Instead of being able to use local files directly out of the repo, instead you'll need to run a step to run a Webpack dev server and load any overlays from there.

Here's my motivations for this:

* One thing we can't do now is to use ESM modules from local `file://` pages.  This means that any js files in node can use `require` but html has to include them with `<script>` in a very fragile way.  Moving to Webpack will let us use CJS or ESM modules in both node and web and will make dependencies a lot cleaner.  This will also make it possible to write a log splitting html front-end to the util/ scripts.  Right now, using any files that depend on other files in both the browser and in node is [very hacky](https://github.com/quisquous/cactbot/blob/main/resources/netregexes.js#L74-L79)

* ngld wants to merge cactbot's plugin code into OverlayPlugin.  One of the things that ngld has mentioned as wanting to solve is the data file loading (e.g. for raidboss triggers and timelines, and oopsy).  This is also filed as #1767.  We can use Webpack to bundle all of these files together locally instead of asking the C# plugin to fetch files from a manifest at runtime.  This will also let raid emulator work without ACT open.  Minor point here also that this means we won't need to list anything in a manifest file anymore.

* I'd also like to move to TypeScript in the future.  We can do this incrementally, and Webpack can help here to handle transforming ts files into js files.  TypeScript will help with catching errors at compile time and providing better Intellisense in VSCode.

I'm mostly filing this issue as a notice for cactbot developers.

**Comment by quisquous** [comment]
@KattTails @mooondark @ShadyWhite @VanSlanzar @Echoring @MaikoTan @Akurosia @JLGarber @panicstevenson @Jaehyuk-Lee @Una-p @valarnin @ngld 

(Pinging folks I see in the last 30 days of commits plus a few others.  Sorry for the ping, just wanted to make sure y'all saw this, as it will be disruptive to cactbot development workflow.)

The timeline for this is before 5.4 drops and probably after Bozja stuff gets added.  So, probably weeks and not months.  I will make sure that there is documentation and time to try this out before merging.

**Comment by quisquous** [comment]
I should also add that this wouldn't be possible without a huge amount of work from @panicstevenson.  (Thank you!)

**Comment by Jaehyuk-Lee** [comment]
I don't disagree as developer, but I feel sorry if this change will make harder to customize things, especially changing trigger alarm text or timeline. There's quite a few users that change timeline or trigger themselves directly to their `raidname.js` or `raidname.txt` to input their tactic.
I guess timeline text file will still be editable after this change, but js trigger file, too?

**Comment by JLGarber** [comment]
I think I have the same concern as Jahyuk-Lee here. What's going to happen to the `user` folder? I don't really care too much about an extra build step, that's just going to slow my local testing down slightly, and I can simply add an extra commit hook to deal with that aspect of it. But asking users to pack their own modifications seems like a heavy lift.

**Comment by quisquous** [comment]
> I think I have the same concern as Jahyuk-Lee here. What's going to happen to the `user` folder? I don't really care too much about an extra build step, that's just going to slow my local testing down slightly, and I can simply add an extra commit hook to deal with that aspect of it. But asking users to pack their own modifications seems like a heavy lift.

This won't change anything for users.  The `user` folder will continue working as-is without modifications, and will continue to load files directly without any bundling step.  Agreed strongly that this would be too much to ask users to do.

It makes me sad that users are changing files directly, because it means every update they have to update it themselves manually.  I personally think that everybody should switch over to using files in `user` instead, and have been trying to encourage people to do that.  I'll make a note about mentioning this in the next release, as well.

I think people modifying local files will already be an issue in the future, as when cactbot plugin merges with overlayplugin, I expect that most people will need to switch to using the http://quisquous.github.io/cactbot/ hosted version, since there will not be a clear way to do updates of local files any more.

Do you think there are things that can be done that would encourage people to use `user` more? I proposed #1786 as one thing.    #860 is another that might encourage more sharing of files inside of `user`.

**Comment by JLGarber** [comment]
> This won't change anything for users. The user folder will continue working as-is without modifications, and will continue to load files directly without any bundling step. Agreed strongly that this would be too much to ask users to do.

Great. Then consider my concerns there covered.

As for the rest of it, I mean, I know for a fact that the breaking change with Overlay 0.3.4 was what made a ton of users *finally* move off of it and onto using NGLD's fork. I think the same thing would happen with modifications directly to the trigger files. Just be Very Clear on each release, starting now, that "direct editing of the trigger files WILL NOT WORK after some undetermined point". If people 100% just Can't Do The Wrong Thing, they will either give up or do the right thing. (Or another Very Wrong Thing, but there are only so many signposts we can put up for them.)

Maybe it would be helpful to have an explicit "user trigger guide" that goes through adding a trigger to `user.js`? The timeline generation guide is super-detailed, and I think it might be necessary to do that here too. ([This was one of the reasons I asked to re-write the main trigger readme in Markdown](https://github.com/quisquous/cactbot/pull/559).) I think moving everything documentation-related out of `user.js` and turning it into a proper readme might help. Or was that done already and I missed it?

**Comment by Jaehyuk-Lee** [comment]
customizing Trigger, most users don't have programming knowledge, they just search text they saw on screen, then change that to some other text. because that's easy for them. They just don't learn how to use `user` folder since there's another easy way.
(plus, there's another reason, "haerubot" updates cactbot by removing entire cactbot directory then download latest cactbot release. So, there's no difference that they need to copy and paste the changed file again. This might be solvable working with haeru.)

customizing Timeline, is there any way to change timeline? Like
`510.0 "Universal Manipulation" sync /:Ascian Prime's Shade:47C8:/` to
`510.0 "Universal Manipulation use A skill" sync /:Ascian Prime's Shade:47C8:/`
To prepare to use A skill. I saw someone doing this.

**Comment by Jaehyuk-Lee** [comment]
Is timeline text file readable? I saw some people using cactbot's timeline text file to apply to SPESPE timeline.

**Comment by quisquous** [comment]
> customizing Trigger, most users don't have programming knowledge, they just search text they saw on screen, then change that to some other text. because that's easy for them. They just don't learn how to use `user` folder since there's another easy way.

Please don't feel like I am ignoring these concerns.  I have similar concerns too about users who are doing something they shouldn't because it's the easiest path.  I think that means to me that there's some work to be done to make what I would like them to do easier.

> (plus, there's another reason, "haerubot" updates cactbot by removing entire cactbot directory then download latest cactbot release. So, there's no difference that they need to copy and paste the changed file again. This might be solvable working with haeru.)

Users can put the user folder somewhere else and point to it with a cactbot option in the config tool.  So, that's probably the best path forward here, rather than using the `cactbot/user/` folder directly if it's getting overwritten.  That's also really unfortunate that all of these users have to update everything after every release no matter what they are doing.

> customizing Timeline, is there any way to change timeline?

Yeah, this is a bit clunky right now, but it is possible right now by writing lines like this:
```
hideall "Universal Manipulation"
510.0 "Universal Manip Custom Text"
```

...and then adding them to the timeline via the triggers file: [example 1](https://github.com/quisquous/cactbot-user/blob/master/raidboss.js#L39-L55) and [example 2](https://github.com/quisquous/cactbot/blob/main/ui/raidboss/data/00-misc/test.js#L10-L36)

I would be happy to try to support a cleaner way to do this if you had ideas.

> Is timeline text file readable? I saw some people using cactbot's timeline text file to apply to SPESPE timeline.

I am not sure what you mean by that.  I am a bit surprised that this would be possible, as I thought spespe timelines were xml, the last time I looked? I suppose you could write a utility to convert them.

If it helps explain better, the timeline text files will continue to exist inside the repository exactly as they are today.  So, if spespe is using them, then that won't break.  However, the text files will be bundled into some raidboss.data.js file via Webpack, and then that js file will be loaded.  So, if a user modifies the original text file, the raidboss.data.js file will not get updated, and the cactbot overlay will use the old unmodified original text file unless the user runs the bundling step themselves to update the raidboss.data.js file with their changes.

**Comment by MaikoTan** [comment]
In fact, most user customize cactbot because they don't like the default alert text. Maybe a GUI for customizing trigger is reasonable?

**Comment by quisquous** [comment]
> In fact, most user customize cactbot because they don't like the default alert text. Maybe a GUI for customizing trigger is reasonable?

Unfortunately, I think it is not easy to do this.  If I had an idea of how to do it, then I would have done it when I added the cactbot config ui.  The issue is that triggers are sometimes JavaScript.  How do you write a GUI for customizing the output of https://github.com/quisquous/cactbot/blob/main/ui/raidboss/data/05-shb/ultimate/the_epic_of_alexander.js#L281-L424?  This is a serious question!  I thought about this a lot, and couldn't come up with any good answers.

If every trigger used "here is an object of language keys and string values", then it would be plausible to add a GUI.  When triggers are JavaScript, it's much less clear how to add a GUI.  Ultimately, I think such a user is going to have to deal with writing JavaScript (or at least editing JavaScript).

Here's what I picture as the best way forward:
* implement cactbot user directories #860
* implement override from #1786 (ideally maybe an override for the top level too and not just every trigger?)

At this point, a user could copy the entire trigger file into their user folder, add override field(s).  Now, they can edit it directly, and ideally chop out the triggers they don't care about.

I think a GUI for editing triggers at that point could look something like this:
* text editor with syntax highlighting
* selection for trigger or zone you want to edit
* it populates a file with that trigger or the whole zone
* errors when you haven't written correct JavaScript, and also basic cactbot testing of triggers (maybe not the full suite, but things like "is this a valid key")
* people can copy that file into their user/ directory

> And, I complain that the current UI organization method is dynamically generated by js and hard to understand. Would it be more "mentain-able"(?) to using React(because I saw you put it in library folder) ?

React is only there because rdmty uses it, and that was imported from elsewhere.

I think I feel the same way about react that I do about jquery.  It's mostly just...it's yet another library to learn and understand how to use.  There's some subjective feeling that I don't think that cactbot ui has enough user interaction or is complicated enough to use it.  At some level, when things are simple enough, there's some feeling on my part that adding additional libraries and abstractions only makes them harder to mentally parse and adds additional maintenance friction for new developers.  I don't mean to say "no libraries, ever", but I want to be careful about when they're needed and which ones to use.

Are there particular parts of the code that you want to change that you find are too complicated?

**Comment by Jaehyuk-Lee** [comment]
Ah, actually timeline using cactbot timeline file might be this, https://github.com/cosmolen/ACT.Timeline
I haven't use any timeline plugins, I don't know details

**Comment by JLGarber** [comment]
> I think I feel the same way about react that I do about jquery. It's mostly just...it's yet another library to learn and understand how to use. There's some subjective feeling that I don't think that cactbot ui has enough user interaction or is complicated enough to use it. At some level, when things are simple enough, there's some feeling on my part that adding additional libraries and abstractions only makes them harder to mentally parse and adds additional maintenance friction for new developers. I don't mean to say "no libraries, ever", but I want to be careful about when they're needed and which ones to use.

To start with, I don't think the position for cactbot should be "we don't want to learn anything new, ever". I'm not advocating for that at all! I think the addition of Typescript would be a good thing. I don't really *like* the addition of Webpack, but I understand why it's going to happen. But I strongly agree with quisquous that if we intend to add other libraries, we need to have a clearly defined problem that A) is solved directly by this library and B) is not practical to solve ourselves with first-party code.

I will freely admit that I don't do a lot of UI work. My focus is usually on how the interface does what it does, rather than how it looks. (That's not a good thing! UI is important, and I should be better about my implementations!)  It's very possible React is a good way to go here. I don't know. But there have already been a [couple](https://github.com/quisquous/cactbot/pull/632) of [times](https://github.com/quisquous/cactbot/issues/1568) where upstream issues in the *build tools* have caused quisquous and panic some scrambling to fix. More third-party code opens the possibility of stuff similar to this.

(I do want to be clear that this is not intended to minimize the opinions of anyone who has reasons we should bring in third-party code. It's not my area of expertise. I just want us to be careful about it.)

**Comment by Jaehyuk-Lee** [comment]
By the way, if trigger and timeline access is harder with this change, I think document translation should be done before. There's a lot users don't understand that documents.
There are some people who can hardly read it, but just click "ok" without reading because they are stressed reading and translating English in their head.

**Comment by Jaehyuk-Lee** [comment]
![K-002](https://user-images.githubusercontent.com/4695371/95711888-a787e800-0c9e-11eb-8635-f2ea3526367c.png)
I thought of these way to GUI customizing triggers. However, this won't possible for function infoTexts.

**Comment by JLGarber** [comment]
I'm still working through a set of triggers for Fractal Hard, but I should be done with that in a day or two. After that, if @quisquous has not already started on it, I will try to write up a proper readme/explanation for how to use the `user/raidboss.js` file, so that users can have that information all in one place rather than trying to read comments. (A lot of people just open it in a basic editor without syntax highlighting, like Notepad, and comments are *very* hard to read without highlighting.) This would not be intended to *replace* the comments in the user file, but rather to expand on them and be a full guide instead of just "helpful hints".

**Comment by MaikoTan** [comment]
In fact, my initial thought is just simply add some mock data to `data` property as new field of trigger object like `mockData` which has all available field of `data` property and `matches` property. When GUI load this trigger, just pass `mockData` to funcs, and you can get pure object of alert text in any language.
But `Nisi` you mentioned seems too complex that even need to add func and complecated objects to `data`, I have no better idea to deal with it.

> The issue is that triggers are sometimes JavaScript.

You means trigger data will still be JavaScript even Cactbot has migrate to TypeScript?

> Here's what I picture as the best way forward:
> * implement cactbot user directories #860
> * implement override from #1786 (ideally maybe an override for the top level too and not just every trigger?)

Seems much better than now. This will permit user overwriting any trigger they don't like without remembering too much.
But it also means when upstream trigger has updated, Cactbot will still load `override` file they customized?
Is it possible to add "This file has been overwriten by user" alert to GUI?

> I think a GUI for editing triggers at that point could look something like this:
> * text editor with syntax highlighting
> * selection for trigger or zone you want to edit
> * it populates a file with that trigger or the whole zone
> * errors when you haven't written correct JavaScript, and also basic cactbot testing of triggers (maybe not the full suite, but things like "is this a valid key")
> * people can copy that file into their user/ directory

Looks great!
May I suggest one more feature about "adding simple trigger(only include regex and infoText without any func) through GUI" ?
 
> Are there particular parts of the code that you want to change that you find are too complicated?
 
Yeah, I means jobs module.
@Echoring and I cost a long time to understand how it work and where to add new code.
Many dynamic parts of jobs module is similar to React, but it will be clearer in React. I don't know how many people here know how to use React though many know around me. So I suggest this to find out is there anyone support.

( @Echoring helped me to translate my reply to English, Thank you! )

**Comment by MaikoTan** [comment]
> To start with, I don't think the position for cactbot should be "we don't want to learn anything new, ever". I'm not advocating for that at all! I think the addition of Typescript would be a good thing. I don't really like the addition of Webpack, but I understand why it's going to happen. But I strongly agree with quisquous that if we intend to add other libraries, we need to have a clearly defined problem that A) is solved directly by this library and B) is not practical to solve ourselves with first-party code.

I also support this, but quisquous seems care about when we add more third-party library, compatibility of old code will be challenged.
As for Webpack and React, I'm sure that they will prevent us from observe result of code change immediately in later developing, but npm can integrate tools to make it easy as no more than 1-2 lines command. (And I remember that there is `watch` command in Webpack let us could build code aitomatically when code changes ?)
As for me, I use VSCode a lot and have successfully made one-key-auto build for a lot of works by `Task` feature. I'm not sure how other editor make this happen, but I believe that they have similar function.

( And thank to Echoring for her excellent translation again! )

**Comment by ngld** [comment]
Sorry for the wall of text... I suppose I have strong opinions on JS dev and automation.

I'd be happy to see Cactbot use Webpack. Imports would make references between files much clearer than the long `<script>` list in each `.html` file. Modern JS features (which you can get through Babel/TypeScript) make life much easier in my opinion. It also makes loading Cactbot over GitHub Pages much faster since it'll significantly reduce the amount of HTTP request needed. If anyone's concerned about Webpack slowing down development, I'd say it does the opposite: If you use `webpack-dev-server`, it'll automatically rebuild the JS bundle whenever a file changes (takes less than a second) and it'll even handle reloads for you. If you implement HMR, you can even keep the current state through the reload.

Regarding the confusion around timelines and triggers: You could just stop shipping the source code since it wouldn't be used anyway. Just ship the bundle files Webpack generates. If people want to modify the source, they should use Git in the first place to manage their changes across updates. At that point adding an additional bundle step to the `git merge/rebase` steps shouldn't make much of a difference.

Making trigger text configurable could be fairly easy IMO. If you add a new property to triggers with a list of templates with placeholders (`Get Final Nisi (?)`, `Pass {nisi} Nisi`, `Pass {nisi} Nisi to {player}`, ...), the JS trigger could then just pick a string from that list (or map) and fill in the placeholder. The config UI could then offer options to override the text for simple triggers and a more complicated interface for these triggers where it'd show a list of templates and allow the user to modify/override them.

Even with Webpack, I'd keep the manifest. I've seen lots of arguments around automatically discovering source files in folders (see C/C++ build tools like CMake, GN, Meson, etc.) and I'm convinced that it's better to explicitly list any dependencies you have (raidboss would depend on the manifest and the manifest depends on the individual timeline/trigger files). This simplifies the integration in Webpack (just a simple loader that converts the manifest into a list of imports) and it makes watching source files easier. If you want to rebuild the bundle whenever a new timeline is added without the manifest, you'd have to watch the whole folder for new files. If you have a manifest, you can instead watch that as any new timeline/trigger file would have to be added.

Regarding the previous issues with npm: The two issues I'm aware of were caused by badly configured environments, not npm itself. The issue with PowerShell was that the user told npm to use PowerShell to process the `package.json` `script` commands which doesn't work since PowerShell doesn't support the syntax.
The other issue where a script file wasn't found was caused by Git-Bash. I generally try to avoid Git-Bash since it's a wonderful source of headaches. Git4Win uses MSYS2 as their base which is essentially a Cygwin fork. The Git-Bash environment uses Cygwin to emulate certain POSIX behaviors and thus will behave differently than cmd. You generally shouldn't run "normal" tools their since this POSIX environment will confuse them. npm detects the Cygwin environment and tries to break out into the normal Windows environment to avoid Cygwin issues. However, to do that it has to convert the Cygwin path `/c/...` to a Windows path. You can do that with the builtin `cygpath` tool in Cygwin but Git4Win removed that tool which causes that step to fail. The end result is that the npm bash script (which you launched from Git-Bash) successfully calls `node.exe` but can't pass a valid path since it can't get it (Git-Bash only supplies to Unix path and the conversion with `cygpath` fails).
If you run the same command from a sane environment (cmd.exe, Conemu, Cygwin, MSYS2, etc.), everything works fine. As a side note, running Windows tools inside a POSIX terminal (like Git4Win's Git-Bash) has other awful side effects and should be avoided as much as possible (Windows tools expect the Windows console API not a POSIX TTY).

**Comment by valarnin** [comment]
I have no complaints here. I use webpack day-to-day already so there's no adjustment for me regarding that.

**Comment by quisquous** [comment]
I'll mention this here explicitly, but if you have time to test out running with Webpack, please try out or review Panic's pull request in #1883 if you can.  It'd be nice to just get a little bit more testing before exposing everybody else to it and finding surprising bugs that our very extensive test suite didn't catch.

The other issue that came up while I was thinking is that converting to Webpack will cause a lot of issues for the ~2 developers running downstream projects that depend on cactbot, e.g. my [https://github.com/quisquous/cactbot-chat](cactbot-chat), but I know there's a few other people.  I think my conclusion for how to support these people is to convert cactbot code into ESM modules that are web-compatible.  Then, downstream users can import those as modules.  It'll prevent them from running anything from `file://`, and so they'll need some form of local webserver / github hosting / etc.  However, it will let them continue using cactbot as an html import with looser coupling.  I think when typescript comes up, we can sort out what to do at that point.  It may be that we bundle everything and also convert ts to a separate js copy and host that as well.


Replying to everybody in turn:

> By the way, if trigger and timeline access is harder with this change, I think document translation should be done before. There's a lot users don't understand that documents or too stressful to understand English.

That's a really good point.

@JLGarber was mentioning that probably the user file documentation needs to be improved anyway.  Maybe the path here can be add user folder, add overrides, improve documentation, get documentation translated, then merge webpack stuff? I will probably do a release shortly, and I would probably start warning people about this so people can get started now.

> Re: trigger customization:
>
> Making trigger text configurable could be fairly easy IMO. If you add a new property to triggers with a list of templates with placeholders (`Get Final Nisi (?)`, `Pass {nisi} Nisi`, `Pass {nisi} Nisi to {player}`, ...), the JS trigger could then just pick a string from that list (or map) and fill in the placeholder. The config UI could then offer options to override the text for simple triggers and a more complicated interface for these triggers where it'd show a list of templates and allow the user to modify/override them.

I think ngld's suggestion above is probably the best approach for the future.  I think ultimately it's not fair to make users write valid JavaScript in order to customize cactbot.  I think probably we will still need the user folder for advanced customization that will never make sense to add to a GUI, but to cover the "rename output text" case, I think the approach of breaking out all of the output of a trigger into template strings, and then returning them would be a great approach.  We'd also have to figure out how to make this work for responses as well.

The downside is that it also means refactoring all of the triggers, and there's a lot of triggers.  So, I don't know that I'd want to block Webpack on doing this work.  I will file an issue with this idea where we can flesh this out in case anybody is interested in working out it in parallel.

> > To start with, I don't think the position for cactbot should be "we don't want to learn anything new, ever". I'm not advocating for that at all! I think the addition of Typescript would be a good thing. I don't really like the addition of Webpack, but I understand why it's going to happen. But I strongly agree with quisquous that if we intend to add other libraries, we need to have a clearly defined problem that A) is solved directly by this library and B) is not practical to solve ourselves with first-party code.
> 
> I also support this, but quisquous seems care about when we add more third-party library, compatibility of old code will be challenged.
> As for Webpack and React, I'm sure that they will prevent us from observe result of code change immediately in later developing, but npm can integrate tools to make it easy as no more than 1-2 lines command. (And I remember that there is `watch` command in Webpack let us could build code aitomatically when code changes ?)

For what it's worth, I don't really like the addition of Webpack either! I knew it was an option that would solve a number of problems and I have resisted adding it for literally years.  I'm reluctant to add additional friction to the project in the form of build steps or conceptual requirements that you have to understand to participate in the project.

I think this is my objection to new things.  It's not that new things are bad, but my position is to be skeptical and reluctant about adding new dependencies.  If we add jquery, then you need to understand how jquery works.  If you add React, then you have to understand React.  The more dependencies and indirection you have, the more a developer needs to hold in their head to understand the entire system.  So, it's an argument for simplicity and resistance to complexity.

At any rate, sorry for the digression.  I think this React discussion is a getting little bit off-topic for Webpack.  I'm happy to discuss in another issue if you'd like to further.

Regarding Webpack, there is a dev server.  In Panic's #1883 patch, you can run `npm run start` to start it, and it will serve the files and automatically rebuild when anything changes.

> Regarding the confusion around timelines and triggers: You could just stop shipping the source code since it wouldn't be used anyway. Just ship the bundle files Webpack generates. If people want to modify the source, they should use Git in the first place to manage their changes across updates. At that point adding an additional bundle step to the `git merge/rebase` steps shouldn't make much of a difference.

I think ultimately the issue is that users aren't developers.  It feels already not ok to ask people to write valid JavaScript.  Asking them to use git and handle merge conflicts across time seems like even more.  I suspect if all users could do this, we wouldn't have ever had questions about customization.  I think your idea above about template scripts is a really good one to handle the common use case of trigger output, and should solve most of this.

> Even with Webpack, I'd keep the manifest. If you want to rebuild the bundle whenever a new timeline is added without the manifest, you'd have to watch the whole folder for new files. If you have a manifest, you can instead watch that as any new timeline/trigger file would have to be added.

Thanks for the advice.  What's the reasoning behind this feeling? Is there a problem with watching folders vs files? Or, is this a preference for explicit over implicit?

> If you run the same command from a sane environment (cmd.exe, Conemu, Cygwin, MSYS2, etc.), everything works fine. As a side note, running Windows tools inside a POSIX terminal (like Git4Win's Git-Bash) has other awful side effects and should be avoided as much as possible (Windows tools expect the Windows console API not a POSIX TTY).

Thanks for the notice.  I didn't realize `git-bash` caused so many issues.

**Comment by Jaehyuk-Lee** [comment]
I didn't though this change will be applied in weeks, I must hurry for document translation.
I'd like to translate [AdvancedCactbot.md](https://github.com/quisquous/cactbot/blob/main/docs/AdvancedCactbot.md#user-folder-config-overrides) especially for `User Folder Config Overrides`. How do I add a new file for translations here?`cactbot/docs/ko/AdvancedCactbot.md`?

**Comment by quisquous** [comment]
> I didn't though this change will be applied in weeks, I must hurry for document translation.
> I'd like to translate [AdvancedCactbot.md](https://github.com/quisquous/cactbot/blob/main/docs/AdvancedCactbot.md#user-folder-config-overrides) especially for `User Folder Config Overrides`. How do I add a new file for translations here?`cactbot/docs/ko/AdvancedCactbot.md`?

I think it might make sense to wait a tiny bit longer until user folders and overrides are added.  I think that will make changing triggers a lot easier, and some of that document might get rewritten.  I am going to try to look into getting that work done today (or this week).

That said, I think a `/ko/` subfolder like you suggest is a good place to put it.

**Comment by quisquous** [comment]
Since I mentioned user folders earlier, I left a comment on https://github.com/quisquous/cactbot/issues/860#issuecomment-711371026, where I'm not sure about shipping user folders prior to Webpack.  I think it will create some backwards compatibility issues if we want to make trigger files into module files into the future.

I am a bit torn, but I think we could ship Webpack without user folders if we had overrides from #1786 and some improved documentation.

**Comment by I-Am-Buggy** [comment]
Looks like this thread has been mostly for contributors, but seemed like a good place to clear up this question so I don't open a new question/issue.

With this change, are there no expected workarounds to directly edit the timeline .txt file and the triggers .js file on a per fight basis?  I ask this because my raid group has been using highly customized timeline and trigger files for savage fights for quite some time. Yes, we're crazy and I maintain a copy of our edits to put the files back after every update.  Probably not ideal, but I don't understand github enough to not do it that way.  I did try the user folder for these changes but I couldn't seem to get those changes to work in the user folder...only general plugin config and css files seem to be working. Is that expected behavior or do I need to tuck e6s.txt and e6s.js in specific subfolders under the user directory to get them to work?  Should we assume this customizability is going away in the near future?

**Comment by quisquous** [comment]
> With this change, are there no expected workarounds to directly edit the timeline .txt file and the triggers .js file on a per fight basis?

I just added https://github.com/quisquous/cactbot/blob/main/docs/CactbotCustomization.md#overriding-raidboss-timelines and https://github.com/quisquous/cactbot/blob/main/docs/CactbotCustomization.md#overriding-raidboss-triggers, which hopefully addresses what you need.

> Yes, we're crazy and I maintain a copy of our edits to put the files back after every update.

:sob:

**Comment by rsod** [comment]
Easier way to modify cactbot is needed. As a right now it's pretty easy to open desired .js file and quickly change what I want to change. With user directory I'd need to go through creating new files, writing new code, making sure that everything works correctly (and I can do so only when I go to raid)... it's just complicated, it's easier to modify code each update, really. To really put this in perspective, once cactbot moves to download stuff from fixed url, it will be easier for me personally to just simply setup a proxy that will intercept request to github and will instead redirect it to locally hosted web server than messing around with these overrides.

**Comment by panicstevenson** [comment]
> With user directory I'd need to go through creating new files, writing new code, making sure that everything works correctly (and I can do so only when I go to raid)... it's just complicated, it's easier to modify code each update, really. 

Maybe I'm confused as to the use-case here, but if you've already updated the triggers, why couldn't you just copy and paste them into `user/raidboss.js` under `Options.Triggers = []`? I guess I don't see why there's a desire to copy and paste the code every update instead of one time?

**Comment by quisquous** [comment]
I think I've done all of the "make it easier to customize cactbot" things that I want to do so far.

The changes to [change trigger text in the UI](https://github.com/quisquous/cactbot/blob/main/docs/CactbotCustomization.md#changing-trigger-text-with-the-cactbot-ui) is the final piece that I think should answer most of what people were editing files for.  Thanks to ngld for the great suggestion on how to do this and to panic for reviewing me changing nearly every trigger to support this.

I think user folders need to wait until after webpack for the reasons I left in a [comment on that issue](https://github.com/quisquous/cactbot/issues/860#issuecomment-711371026).

In summary, I think I would like to try to get Webpack turned on and landed in the next week or so, which would give plenty of time to flush out any issues before the next raid tier drops.

**Comment by quisquous** [comment]
I haven't found anything in initial testing so far.  Has anybody else but panicstevenson and I tested this patch?

Because I hate having to remember to run console commands, I wrote up https://github.com/quisquous/npm-auto-start-plugin/releases/tag/v0.0.1, which other folks might be interested in.  It's an OverlayPlugin plugin that starts the webpack dev server when ACT starts and kills it when ACT ends.  It's not the fanciest thing, but it removes overhead for me.

Still planning to land this webpack pr in the next week.
